viewing_party/party.py

```diff
@@ -1,23 +1,171 @@
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
-        
+
+def get_unique_watched(user_data):
+    """Return movies the user has watched that their friends have not watched."""
+    user_movies_watched = user_data["watched"]  # list of dicts of movies
+    user_movie_titles = [
+        movie["title"].lower() for movie in user_movies_watched
+    ]  # list of titles only
+
+    friends_movies_watched = []  # list of dicts of movies (all friends combined)
+    for friend in user_data["friends"]:
+        friends_movies_watched.extend(friend["watched"])
+    friend_movie_titles = [
+        movie["title"].lower() for movie in friends_movies_watched
+    ]  # list of titles only
+
+    user_unique_movie_titles = set(user_movie_titles) - set(friend_movie_titles)
+    user_unique_movies = []
+
+    for movie in user_movies_watched:
+        if movie["title"].lower() in user_unique_movie_titles:
+            user_unique_movies.append(movie)
+
+    return user_unique_movies
+
+
+def get_friends_unique_watched(user_data):
+    """Return movies the user's friends have watched that the user has not watched."""
+    user_movies_watched = user_data["watched"]  # list of dicts of movies
+    user_movie_titles = [
+        movie["title"].lower() for movie in user_movies_watched
+    ]  # list of titles only
+
+    friends_movies_watched = []  # list of dicts of movies (all friends combined)
+    for friend in user_data["friends"]:
+        friends_movies_watched.extend(friend["watched"])
+    friend_movie_titles = [
+        movie["title"].lower() for movie in friends_movies_watched
+    ]  # list of titles only
+
+    friend_unique_movie_titles = set(friend_movie_titles) - set(user_movie_titles)
+    friend_unique_movies = []
+
+    for movie in friends_movies_watched:
+        if (
+            movie["title"].lower() in friend_unique_movie_titles
+            and movie not in friend_unique_movies
+        ):
+            friend_unique_movies.append(movie)
+
+    return friend_unique_movies
+
+
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
 
+
+def get_available_recs(user_data):
+    """Return recommended movies that the user subscribes to."""
+    recommended_movies = []
+    friends_unique = get_friends_unique_watched(user_data)
+
+    for movie in friends_unique:
+        if movie["host"] in user_data["subscriptions"]:
+            recommended_movies.append(movie)
+
+    return recommended_movies
+
+
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
 
+
+def get_new_rec_by_genre(user_data):
+    """Return recommended movies of the user's most frequently watched genre."""
+    most_watched_genre = get_most_watched_genre(user_data)
+    movies_not_watched = get_friends_unique_watched(user_data)
+
+    recommended_by_genre = []

```

Comment:
[naming] I really appreciate the names you've used; they make it easy to follow.